design-docs/Scalar adapter configuration.md

```diff
@@ -0,0 +1,195 @@
+val customScalarsMapping: MapProperty<String, String>
+```
+
+Used to generate the wanted type in generated models / adapters (`kotlin.Any` is used by default when no mapping is
+specified).
+
+A “Type” class is also generated.
+
+Example:
+
+```kotlin
+public class MyDate {
+  public companion object {
+    public val type: CustomScalarType = CustomScalarType("MyDate", "com.example.MyDate")
+  }
+}
+```
+
+In the generated `ResponseAdapter` and `VariablesAdapter` objects, the registered scalar adapter is found in
+the `customScalarAdapters` parameter, like so:
+
+```kotlin
+public override fun fromJson(reader: JsonReader, customScalarAdapters: CustomScalarAdapters):
+    GetDateQuery.Data {
+// ...
+  3 -> date = customScalarAdapters.responseAdapterFor<MyDate>(com.example.type.MyDate.type).nullable().fromJson(reader, customScalarAdapters)
+// ...
+```
+
+(`AnyAdapter` is used by default when no mapping is declared).
+
+Whereas, for **built-in** scalars (e.g. `ID` and `String`), the type and adapter to use are hard-coded in the generated
+code:
+
+```kotlin
+2 -> id = StringAdapter.fromJson(reader, customScalarAdapters)
+```
+
+The hardcoded types / adapters for built-in scalars are
+in [`KotlinResolver`](https://github.com/apollographql/apollo-kotlin/blob/main/apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/codegen/kotlin/KotlinResolver.kt#L26)
+and [`JavaResolver`](https://github.com/apollographql/apollo-kotlin/blob/main/apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/codegen/java/JavaResolver.kt)
+.
+
+The `customScalarAdapters` comes from what is configured on the `ApolloClient` at runtime by the user. It is passed to
+requests (in the `executionContext`), and then to the adapters in `HttpNetworkTransport` / `WebSocketNetworkTransport`.
+
+## Proposed changes
+
+1. Add the ability to specify an adapter in addition to the type class name when declaring the scalar
+   mappings in the plugin configuration
+2. Allow declaring mappings for built-in scalars (e.g. `ID`)
+
+### User facing API
+
+Plugin configuration:
+
+```kotlin
+apollo {
+  mapScalar("MyDate", "com.example.MyDate", NoArgConstructorAdapterInitializer("com.example.MyDateAdapter"))
+
+  mapScalar("ID", "kotlin.Long", SingletonAdapterInitializer("com.apollographql.apollo3.api.LongAdapter"))
+  
+  mapScalar("MyLong", "kotlin.Long")
+}
+```
+
+The second parameter is a sealed interface to account of the different ways to use the given adapter name:
+
+```kotlin
+sealed interface AdapterInitializer
+
+// The adapter will be instantiated in the generated code
+class SingletonAdapterInitializer(val qualifiedName: String): AdapterInitializer
+
+// The adapter will be used as-is (it's an object or a public val)
+class NoArgConstructorAdapterInitializer(val qualifiedName: String): AdapterInitializer
+
+// The adapter will be looked up in the `customScalarAdapters` parameter (same as current behavior)
+object RuntimeAdapterInitializer: AdapterInitializer
```

Comment:
I know I'm the one who suggested this initially but I think we should avoid Kotlin APIs in the Gradle plugin API. 

Mainly for Groovy users + because we use R8 to relocate all of the Kotlin-stdlib, this feels dangerous. I'd suggest something like this instead:

```kotlin
class Service {
  fun mapScalar(graphQLName: String, kotlinName: String)
  fun mapScalarUsingKotlinSingleton(graphQLName: String, kotlinName: String, singletonName: String)
  fun mapScalarUsingKotlinConstructor(graphQLName: String, kotlinName: String, constructorName: String)

  fun mapScalarToKotlinLong(graphQLName: String)
  fun mapScalarToKotlinInt(graphQLName: String)
  fun mapScalarToKotlinString(graphQLName: String)
  fun mapScalarToKotlinBoolean(graphQLName: String)
  fun mapScalarToKotlinMap(graphQLName: String)

  fun mapScalarToJavaLong(graphQLName: String)
  fun mapScalarToJavaInteger(graphQLName: String)
  fun mapScalarToJavaString(graphQLName: String)
  fun mapScalarToJavaBoolean(graphQLName: String)
  fun mapScalarToJavaMap(graphQLName: String)
}
```

Reply:
Agreed! 
Plus I actually find the syntax easier to use 👍

Shouldn't `mapScalarUsingKotlinSingleton` / `mapScalarUsingKotlinConstructor` be `mapScalarUsingSingleton` / `mapScalarUsingConstructor`?

Reply:
> Shouldn't mapScalarUsingKotlinSingleton / mapScalarUsingKotlinConstructor be mapScalarUsingSingleton / mapScalarUsingConstructor?

I was wondering the same. `Singleton` is a Kotlin concept so it will not work with Java but maybe we can throw a nice error if that happens?

Reply:
I was assuming this would work also in Java with static fields:
```java
public class MyAdapters {
  public static final Adapter<MyDate> DATE_ADAPTER = new Adapter<MyDate>() {...}
}
```

```kotlin
mapScalarUsingSingleton("Date", "com.example.MyDate", "com.example.MyAdapters.DATE_ADAPTER")
```

But now I'm realizing maybe the name "singleton" is not the most accurate?

Reply:
I think I agree, the distinction is not that useful. The documentation can explain how it works, and the fact that an instance or an instantiation expression can both be used.

In that case, maybe we can keep the name `mapScalar`?

```kotlin
  /**
   * Map a GraphQL scalar type to the Java/Kotlin type and adapter configured at runtime.
   * 
   * For example:
   * `mapScalar("Date", "com.example.Date")`
   */
  fun mapScalar(graphQLName: String, targetName: String)

  /**
   * Map a GraphQL scalar type to the Java/Kotlin type and provided adapter expression.
   *
   * For example:
   * `mapScalar("Date", "com.example.Date", "com.example.DateAdapter")` (an instance property or object)
   * `mapScalar("Date", "com.example.Date", "com.example.DateAdapter()")` (instantiate the class on the fly)
   */
  fun mapScalar(graphQLName: String, targetName: String, adatpterExpression: String)

```

Reply:
`mapScalarToKotlinMap` / `mapScalarToJavaMap` should be `mapScalarToKotlinAny` / `mapScalarToJavaObject` or am I missing something?

Reply:
Yes, indeed. Until we have https://github.com/apollographql/apollo-kotlin/issues/3243, we can't use Map or other parametrized classes.